enable instantaneous_rate() for lists of spike trains

rgutzen opened the pull request:
Fixes #133 by checking if the input `spiketrain` is of type `list` and calling the function `instantaneous_rate()` for each element of the list. In this case the function returns a list of analog signals.

pep8speaks commented:
Hello @rgutzen! Thanks for submitting the PR.

 - In the file [`elephant/statistics.py`](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py), following are the PEP8 issues :

> [Line 3:80](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L3): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 79 characters)
> [Line 174:80](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L174): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 255:74](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L255): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 293:73](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L293): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 294:75](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L294): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 317:20](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L317): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 336:20](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L336): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 343:60](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L343): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 429:22](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L429): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 509:24](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L509): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 510:24](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L510): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 510:50](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L510): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 647:34](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L647): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 648:34](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L648): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 675:74](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L675): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 677:73](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L677): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 681:73](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L681): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 739:80](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L739): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 740:80](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L740): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 79 characters)
> [Line 755:80](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L755): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 792:80](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L792): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 824:80](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L824): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 945:34](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L945): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 946:34](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L946): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 1086:16](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L1086): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 1100:80](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L1100): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 1117:80](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L1117): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 1135:80](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L1135): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 79 characters)
> [Line 1181:15](https://github.com/NeuralEnsemble/elephant/blob/777aed6e716a3a9f0c14e78e254d7deb0945f3ee/elephant/statistics.py#L1181): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent

alperyeg commented:
@rgutzen that looks good. Can you provide a test which checks for a list of 2 spiketrains?

rgutzen commented:
- Changed the output of the instantaneous_rate function (when given a list of spike trains) from a list of analog signals to a single analog signal representing the estimated population rate of the spike train list.
- Added a corresponding unit test based on the property that `instantaneous_rate(spiketrain_1) + instantaneous_rate(spiketrain_2) == instantaneous_rate([spiketrain_1, spiketrain_2])` when the kernel is fixed.

alperyeg commented on elephant/statistics.py:
```diff
@@ -732,14 +732,33 @@ def instantaneous_rate(spiketrain, sampling_period, kernel='auto',
-                                      cutoff=cutoff, t_start=t_start, t_stop=t_start, trim=trim)
-            rates += [rate]
-        return rates
+            if not isinstance(st, SpikeTrain):
```
All the consistency check should go into a private function for easier handling.
